homeassistant/components/hydrawise/sensor.py

```diff
@@ -44,23 +48,21 @@ def update(self):
         """Get the latest data and updates the states."""
         mydata = self.hass.data[DATA_HYDRAWISE].data
         _LOGGER.debug("Updating Hydrawise sensor: %s", self._name)
+        relay_data = mydata.relays[self.data["relay"] - 1]
         if self._sensor_type == "watering_time":
-            if not mydata.running:
-                self._state = 0
+            if relay_data["timestr"] == "Now":
+                self._state = int(relay_data["run"] / 60)
             else:
-                if int(mydata.running[0]["relay"]) == self.data["relay"]:
-                    self._state = int(mydata.running[0]["time_left"] / 60)
-                else:
-                    self._state = 0
+                self._state = 0
         else:  # _sensor_type == 'next_cycle'
-            for relay in mydata.relays:
-                if relay["relay"] == self.data["relay"]:
-                    if relay["nicetime"] == "Not scheduled":
-                        self._state = "not_scheduled"
-                    else:
-                        self._state = f"{relay['nicetime'].split(',')[0]} {relay['nicetime'].split(' ')[3]}"
+            next_cycle = min(relay_data["time"], TWO_YEAR_SECONDS)
+            _LOGGER.debug("New cycle time: %s", next_cycle)
+            self._state = dt.as_local(
+                dt.utc_from_timestamp(dt.as_timestamp(dt.now()) + next_cycle)
+            ).isoformat()
 
     @property
     def icon(self):
-        """Icon to use in the frontend, if any."""
-        return DEVICE_MAP[self._sensor_type][DEVICE_MAP_INDEX.index("ICON_INDEX")]
+        """Return the icon to use in the frontend, if any."""
+        if self._sensor_type == "watering_time":
```

Comment:
It's better to use the `DEVICE_MAP` to set the icon so we have all sensor type settings in one place.

Reply:
OK. I put this back in the DEVICE_MAP.